transcom/trdm-lambda
# B 20145 automate tests

## [Ticket B-20145](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-20145) 

> [!TIP]
> This is valuable information for issues that will be encountered when we decide to automate 508 compliance and attempt local testing on the mymove repo

> [!NOTE]  
> The following discussion highlights limitations within the local environment. These issues DO NOT replicate in the live CircleCI environment. Please see more information in the relevant articles section.

# Summary

This reintroduces maven testing for the trdm pipeline by containerizing the CircleCI process, the milmove dependencies, and the postgres container for the test db.

A three container docker-in-docker approach was utilized due to known CircleCI CLI issues and limitations for local development preventing parallel images. These containers will use a remote docker setup with the CircleCI deprecated version 20. This parallel image error can easily be replicated by trying to run one of the mymove CircleCI jobs locally that use an executor with two images. 

The entire testing workflow was also put into a single job step due as workflows are another limitation that cannot run locally. By putting it all into a single job step, we include all dependencies meaning we can properly test in our local environments. The testing containers rely on the main mymove branch and execute migrations accordingly.

Workspaces and Docker volumes were not utilized due to additional restrictions encountered with host machine access, a workaround of input streaming was put in place to achieve the same functionality.

# How to test locally
-	Make sure CircleCI CLI is installed
-	Make sure Docker is running (Version doesn’t matter due to remote docker versioning)
-	Within the repository, run `circleci local execute test`

# Relevant Articles
Explaining the exact issue encountered
https://discuss.circleci.com/t/circleci-cli-error-response-from-daemon-invalid-uts-mode/48081

Open discussion detailing the parallel image
https://github.com/CircleCI-Public/circleci-cli/issues/970

Discussion detailing the last working docker version
https://github.com/andyatkinson/rideshare/issues/99

Marking 20.10.24 as the deprecated version 
https://circleci.com/docs/building-docker-images/#docker-version

Known CircleCI CLI local limitations
https://circleci.com/docs/how-to-use-the-circleci-local-cli/#limitations-of-running-jobs-locally
